From 5ff051f936a5bc4f5e1edb17ee1c3149b66644a2 Mon Sep 17 00:00:00 2001
From: Mark Reynolds <mreynolds@redhat.com>
Date: Wed, 16 Jul 2025 20:54:48 -0400
Subject: [PATCH] Issue 6888 - Missing access JSON logging for TLS/Client auth

Description:

TLS/Client auth logging was not converted to JSON (auth.c got missed)

Relates: https://github.com/389ds/389-ds-base/issues/6888

Reviewed by: spichugi(Thanks!)
---
 .../logging/access_json_logging_test.py       |  96 ++++++++-
 ldap/servers/slapd/accesslog.c                | 114 +++++++++++
 ldap/servers/slapd/auth.c                     | 182 +++++++++++++-----
 ldap/servers/slapd/log.c                      |   2 +
 ldap/servers/slapd/slapi-private.h            |  10 +
 5 files changed, 353 insertions(+), 51 deletions(-)

diff --git a/dirsrvtests/tests/suites/logging/access_json_logging_test.py b/dirsrvtests/tests/suites/logging/access_json_logging_test.py
index ae91dc487..f0dc861a7 100644
--- a/dirsrvtests/tests/suites/logging/access_json_logging_test.py
+++ b/dirsrvtests/tests/suites/logging/access_json_logging_test.py
@@ -19,6 +19,8 @@ from lib389.idm.user import UserAccounts
 from lib389.dirsrv_log import DirsrvAccessJSONLog
 from lib389.index import VLVSearch, VLVIndex
 from lib389.tasks import Tasks
+from lib389.config import CertmapLegacy
+from lib389.nss_ssl import NssSsl
 from ldap.controls.vlv import VLVRequestControl
 from ldap.controls.sss import SSSRequestControl
 from ldap.controls import SimplePagedResultsControl
@@ -67,11 +69,11 @@ def get_log_event(inst, op, key=None, val=None, key2=None, val2=None):
                     if val == str(event[key]).lower() and \
                        val2 == str(event[key2]).lower():
                         return event
-
-            elif key is not None and key in event:
-                val = str(val).lower()
-                if val == str(event[key]).lower():
-                    return event
+            elif key is not None:
+                if key in event:
+                    val = str(val).lower()
+                    if val == str(event[key]).lower():
+                        return event
             else:
                 return event
 
@@ -163,6 +165,7 @@ def test_access_json_format(topo_m2, setup_test):
         14. Test PAGED SEARCH is logged correctly
         15. Test PERSISTENT SEARCH is logged correctly
         16. Test EXTENDED OP
+        17. Test TLS_INFO is logged correctly
     :expectedresults:
         1. Success
         2. Success
@@ -180,6 +183,7 @@ def test_access_json_format(topo_m2, setup_test):
         14. Success
         15. Success
         16. Success
+        17. Success
     """
 
     inst = topo_m2.ms["supplier1"]
@@ -560,6 +564,88 @@ def test_access_json_format(topo_m2, setup_test):
     assert event['oid_name'] == "REPL_END_NSDS50_REPLICATION_REQUEST_OID"
     assert event['name'] == "replication-multisupplier-extop"
 
+    #
+    # TLS INFO/TLS CLIENT INFO
+    #
+    RDN_TEST_USER = 'testuser'
+    RDN_TEST_USER_WRONG = 'testuser_wrong'
+    inst.enable_tls()
+    inst.restart()
+
+    users = UserAccounts(inst, DEFAULT_SUFFIX)
+    user = users.create(properties={
+        'uid': RDN_TEST_USER,
+        'cn': RDN_TEST_USER,
+        'sn': RDN_TEST_USER,
+        'uidNumber': '1000',
+        'gidNumber': '2000',
+        'homeDirectory': f'/home/{RDN_TEST_USER}'
+    })
+
+    ssca_dir = inst.get_ssca_dir()
+    ssca = NssSsl(dbpath=ssca_dir)
+    ssca.create_rsa_user(RDN_TEST_USER)
+    ssca.create_rsa_user(RDN_TEST_USER_WRONG)
+
+    # Get the details of where the key and crt are.
+    tls_locs = ssca.get_rsa_user(RDN_TEST_USER)
+    tls_locs_wrong = ssca.get_rsa_user(RDN_TEST_USER_WRONG)
+
+    user.enroll_certificate(tls_locs['crt_der_path'])
+
+    # Turn on the certmap.
+    cm = CertmapLegacy(inst)
+    certmaps = cm.list()
+    certmaps['default']['DNComps'] = ''
+    certmaps['default']['FilterComps'] = ['cn']
+    certmaps['default']['VerifyCert'] = 'off'
+    cm.set(certmaps)
+
+    # Check that EXTERNAL is listed in supported mechns.
+    assert (inst.rootdse.supports_sasl_external())
+
+    # Restart to allow certmaps to be re-read: Note, we CAN NOT use post_open
+    # here, it breaks on auth. see lib389/__init__.py
+    inst.restart(post_open=False)
+
+    # Attempt a bind with TLS external
+    inst.open(saslmethod='EXTERNAL', connOnly=True, certdir=ssca_dir,
+              userkey=tls_locs['key'], usercert=tls_locs['crt'])
+    inst.restart()
+
+    event = get_log_event(inst, "TLS_INFO")
+    assert event is not None
+    assert 'tls_version' in event
+    assert 'keysize' in event
+    assert 'cipher' in event
+
+    event = get_log_event(inst, "TLS_CLIENT_INFO",
+                          "subject",
+                          "CN=testuser,O=testing,L=389ds,ST=Queensland,C=AU")
+    assert event is not None
+    assert 'tls_version' in event
+    assert 'keysize' in event
+    assert 'issuer' in event
+
+    event = get_log_event(inst, "TLS_CLIENT_INFO",
+                          "client_dn",
+                          "uid=testuser,ou=People,dc=example,dc=com")
+    assert event is not None
+    assert 'tls_version' in event
+    assert event['msg'] == "client bound"
+
+    # Check for failed certmap error
+    with pytest.raises(ldap.INVALID_CREDENTIALS):
+        inst.open(saslmethod='EXTERNAL', connOnly=True, certdir=ssca_dir,
+                  userkey=tls_locs_wrong['key'],
+                  usercert=tls_locs_wrong['crt'])
+
+    event = get_log_event(inst, "TLS_CLIENT_INFO", "err", -185)
+    assert event is not None
+    assert 'tls_version' in event
+    assert event['msg'] == "failed to map client certificate to LDAP DN"
+    assert event['err_msg'] == "Certificate couldn't be mapped to an ldap entry"
+
 
 if __name__ == '__main__':
     # Run isolated
diff --git a/ldap/servers/slapd/accesslog.c b/ldap/servers/slapd/accesslog.c
index 68022fe38..072ace203 100644
--- a/ldap/servers/slapd/accesslog.c
+++ b/ldap/servers/slapd/accesslog.c
@@ -1147,3 +1147,117 @@ slapd_log_access_sort(slapd_log_pblock *logpb)
 
     return rc;
 }
+
+/*
+ * TLS connection
+ *
+ * int32_t log_format
+ * time_t conn_time
+ * uint64_t conn_id
+ * const char *msg
+ * const char *tls_version
+ * int32_t keysize
+ * const char *cipher
+ * int32_t err
+ * const char *err_str
+ */
+int32_t
+slapd_log_access_tls(slapd_log_pblock *logpb)
+{
+    int32_t rc = 0;
+    char *msg = NULL;
+    json_object *json_obj = NULL;
+
+    if ((json_obj = build_base_obj(logpb, "TLS_INFO")) == NULL) {
+        return rc;
+    }
+
+    if (logpb->msg) {
+        json_object_object_add(json_obj, "msg", json_obj_add_str(logpb->msg));
+    }
+    if (logpb->tls_version) {
+        json_object_object_add(json_obj, "tls_version", json_obj_add_str(logpb->tls_version));
+    }
+    if (logpb->cipher) {
+        json_object_object_add(json_obj, "cipher", json_obj_add_str(logpb->cipher));
+    }
+    if (logpb->keysize) {
+        json_object_object_add(json_obj, "keysize", json_object_new_int(logpb->keysize));
+    }
+    if (logpb->err_str) {
+        json_object_object_add(json_obj, "err", json_object_new_int(logpb->err));
+        json_object_object_add(json_obj, "err_msg", json_obj_add_str(logpb->err_str));
+    }
+
+    /* Convert json object to string and log it */
+    msg = (char *)json_object_to_json_string_ext(json_obj, logpb->log_format);
+    rc = slapd_log_access_json(msg);
+
+    /* Done with JSON object - free it */
+    json_object_put(json_obj);
+
+    return rc;
+}
+
+/*
+ * TLS client auth
+ *
+ * int32_t log_format
+ * time_t conn_time
+ * uint64_t conn_id
+ * const char* tls_version
+ * const char* keysize
+ * const char* cipher
+ * const char* msg
+ * const char* subject
+ * const char* issuer
+ * int32_t err
+ * const char* err_str
+ * const char *client_dn
+ */
+int32_t
+slapd_log_access_tls_client_auth(slapd_log_pblock *logpb)
+{
+    int32_t rc = 0;
+    char *msg = NULL;
+    json_object *json_obj = NULL;
+
+    if ((json_obj = build_base_obj(logpb, "TLS_CLIENT_INFO")) == NULL) {
+        return rc;
+    }
+
+    if (logpb->tls_version) {
+        json_object_object_add(json_obj, "tls_version", json_obj_add_str(logpb->tls_version));
+    }
+    if (logpb->cipher) {
+        json_object_object_add(json_obj, "cipher", json_obj_add_str(logpb->cipher));
+    }
+    if (logpb->keysize) {
+        json_object_object_add(json_obj, "keysize", json_object_new_int(logpb->keysize));
+    }
+    if (logpb->subject) {
+        json_object_object_add(json_obj, "subject", json_obj_add_str(logpb->subject));
+    }
+    if (logpb->issuer) {
+        json_object_object_add(json_obj, "issuer", json_obj_add_str(logpb->issuer));
+    }
+    if (logpb->client_dn) {
+        json_object_object_add(json_obj, "client_dn", json_obj_add_str(logpb->client_dn));
+    }
+    if (logpb->msg) {
+        json_object_object_add(json_obj, "msg", json_obj_add_str(logpb->msg));
+    }
+    if (logpb->err_str) {
+        json_object_object_add(json_obj, "err", json_object_new_int(logpb->err));
+        json_object_object_add(json_obj, "err_msg", json_obj_add_str(logpb->err_str));
+    }
+
+    /* Convert json object to string and log it */
+    msg = (char *)json_object_to_json_string_ext(json_obj, logpb->log_format);
+    rc = slapd_log_access_json(msg);
+
+    /* Done with JSON object - free it */
+    json_object_put(json_obj);
+
+    return rc;
+}
diff --git a/ldap/servers/slapd/auth.c b/ldap/servers/slapd/auth.c
index e4231bf45..48e4b7129 100644
--- a/ldap/servers/slapd/auth.c
+++ b/ldap/servers/slapd/auth.c
@@ -1,6 +1,6 @@
 /** BEGIN COPYRIGHT BLOCK
  * Copyright (C) 2001 Sun Microsystems, Inc. Used by permission.
- * Copyright (C) 2005 Red Hat, Inc.
+ * Copyright (C) 2025 Red Hat, Inc.
  * All rights reserved.
  *
  * License: GPL (version 3 or any later version).
@@ -363,19 +363,32 @@ handle_bad_certificate(void *clientData, PRFileDesc *prfd)
     char sbuf[BUFSIZ], ibuf[BUFSIZ];
     Connection *conn = (Connection *)clientData;
     CERTCertificate *clientCert = slapd_ssl_peerCertificate(prfd);
-
     PRErrorCode errorCode = PR_GetError();
     char *subject = subject_of(clientCert);
     char *issuer = issuer_of(clientCert);
-    slapi_log_access(LDAP_DEBUG_STATS,
-                     "conn=%" PRIu64 " " SLAPI_COMPONENT_NAME_NSPR " error %i (%s); unauthenticated client %s; issuer %s\n",
-                     conn->c_connid, errorCode, slapd_pr_strerror(errorCode),
-                     subject ? escape_string(subject, sbuf) : "NULL",
-                     issuer ? escape_string(issuer, ibuf) : "NULL");
+    int32_t log_format = config_get_accesslog_log_format();
+    slapd_log_pblock logpb = {0};
+
+    if (log_format != LOG_FORMAT_DEFAULT) {
+        slapd_log_pblock_init(&logpb, log_format, NULL);
+        logpb.conn_id = conn->c_connid;
+        logpb.msg = "unauthenticated client";
+        logpb.subject = subject ? escape_string(subject, sbuf) : "NULL";
+        logpb.issuer = issuer ? escape_string(issuer, ibuf) : "NULL";
+        logpb.err = errorCode;
+        logpb.err_str = slapd_pr_strerror(errorCode);
+        slapd_log_access_tls_client_auth(&logpb);
+    } else {
+        slapi_log_access(LDAP_DEBUG_STATS,
+                        "conn=%" PRIu64 " " SLAPI_COMPONENT_NAME_NSPR " error %i (%s); unauthenticated client %s; issuer %s\n",
+                        conn->c_connid, errorCode, slapd_pr_strerror(errorCode),
+                        subject ? escape_string(subject, sbuf) : "NULL",
+                        issuer ? escape_string(issuer, ibuf) : "NULL");
+    }
     if (issuer)
-        free(issuer);
+        slapi_ch_free_string(&issuer);
     if (subject)
-        free(subject);
+        slapi_ch_free_string(&subject);
     if (clientCert)
         CERT_DestroyCertificate(clientCert);
     return -1; /* non-zero means reject this certificate */
@@ -394,7 +407,8 @@ handle_handshake_done(PRFileDesc *prfd, void *clientData)
 {
     Connection *conn = (Connection *)clientData;
     CERTCertificate *clientCert = slapd_ssl_peerCertificate(prfd);
-
+    int32_t log_format = config_get_accesslog_log_format();
+    slapd_log_pblock logpb = {0};
     char *clientDN = NULL;
     int keySize = 0;
     char *cipher = NULL;
@@ -403,19 +417,39 @@ handle_handshake_done(PRFileDesc *prfd, void *clientData)
     SSLCipherSuiteInfo cipherInfo;
     char *subject = NULL;
     char sslversion[64];
+    int err = 0;
 
     if ((slapd_ssl_getChannelInfo(prfd, &channelInfo, sizeof(channelInfo))) != SECSuccess) {
         PRErrorCode errorCode = PR_GetError();
-        slapi_log_access(LDAP_DEBUG_STATS,
-                         "conn=%" PRIu64 " SSL failed to obtain channel info; " SLAPI_COMPONENT_NAME_NSPR " error %i (%s)\n",
-                         conn->c_connid, errorCode, slapd_pr_strerror(errorCode));
+        if (log_format != LOG_FORMAT_DEFAULT) {
+            slapd_log_pblock_init(&logpb, log_format, NULL);
+            logpb.conn_id = conn->c_connid;
+            logpb.err = errorCode;
+            logpb.err_str = slapd_pr_strerror(errorCode);
+            logpb.msg = "SSL failed to obtain channel info; " SLAPI_COMPONENT_NAME_NSPR;
+            slapd_log_access_tls(&logpb);
+        } else {
+            slapi_log_access(LDAP_DEBUG_STATS,
+                    "conn=%" PRIu64 " SSL failed to obtain channel info; " SLAPI_COMPONENT_NAME_NSPR " error %i (%s)\n",
+                    conn->c_connid, errorCode, slapd_pr_strerror(errorCode));
+        }
         goto done;
     }
+
     if ((slapd_ssl_getCipherSuiteInfo(channelInfo.cipherSuite, &cipherInfo, sizeof(cipherInfo))) != SECSuccess) {
         PRErrorCode errorCode = PR_GetError();
-        slapi_log_access(LDAP_DEBUG_STATS,
-                         "conn=%" PRIu64 " SSL failed to obtain cipher info; " SLAPI_COMPONENT_NAME_NSPR " error %i (%s)\n",
-                         conn->c_connid, errorCode, slapd_pr_strerror(errorCode));
+        if (log_format != LOG_FORMAT_DEFAULT) {
+            slapd_log_pblock_init(&logpb, log_format, NULL);
+            logpb.conn_id = conn->c_connid;
+            logpb.err = errorCode;
+            logpb.err_str = slapd_pr_strerror(errorCode);
+            logpb.msg = "SSL failed to obtain cipher info; " SLAPI_COMPONENT_NAME_NSPR;
+            slapd_log_access_tls(&logpb);
+        } else {
+            slapi_log_access(LDAP_DEBUG_STATS,
+                    "conn=%" PRIu64 " SSL failed to obtain cipher info; " SLAPI_COMPONENT_NAME_NSPR " error %i (%s)\n",
+                    conn->c_connid, errorCode, slapd_pr_strerror(errorCode));
+        }
         goto done;
     }
 
@@ -434,47 +468,84 @@ handle_handshake_done(PRFileDesc *prfd, void *clientData)
 
     if (config_get_SSLclientAuth() == SLAPD_SSLCLIENTAUTH_OFF) {
         (void)slapi_getSSLVersion_str(channelInfo.protocolVersion, sslversion, sizeof(sslversion));
-        slapi_log_access(LDAP_DEBUG_STATS, "conn=%" PRIu64 " %s %i-bit %s\n",
-                         conn->c_connid,
-                         sslversion, keySize, cipher ? cipher : "NULL");
+        if (log_format != LOG_FORMAT_DEFAULT) {
+            slapd_log_pblock_init(&logpb, log_format, NULL);
+            logpb.conn_id = conn->c_connid;
+            logpb.tls_version = sslversion;
+            logpb.keysize = keySize;
+            logpb.cipher = cipher ? cipher : "NULL";
+            slapd_log_access_tls(&logpb);
+        } else {
+            slapi_log_access(LDAP_DEBUG_STATS, "conn=%" PRIu64 " %s %i-bit %s\n",
+                    conn->c_connid,
+                    sslversion, keySize, cipher ? cipher : "NULL");
+        }
         goto done;
     }
     if (clientCert == NULL) {
         (void)slapi_getSSLVersion_str(channelInfo.protocolVersion, sslversion, sizeof(sslversion));
-        slapi_log_access(LDAP_DEBUG_STATS, "conn=%" PRIu64 " %s %i-bit %s\n",
-                         conn->c_connid,
-                         sslversion, keySize, cipher ? cipher : "NULL");
+        if (log_format != LOG_FORMAT_DEFAULT) {
+            slapd_log_pblock_init(&logpb, log_format, NULL);
+            logpb.conn_id = conn->c_connid;
+            logpb.tls_version = sslversion;
+            logpb.keysize = keySize;
+            logpb.cipher = cipher ? cipher : "NULL";
+            slapd_log_access_tls(&logpb);
+        } else {
+            slapi_log_access(LDAP_DEBUG_STATS, "conn=%" PRIu64 " %s %i-bit %s\n",
+                    conn->c_connid,
+                    sslversion, keySize, cipher ? cipher : "NULL");
+        }
     } else {
         subject = subject_of(clientCert);
         if (!subject) {
             (void)slapi_getSSLVersion_str(channelInfo.protocolVersion,
                                           sslversion, sizeof(sslversion));
-            slapi_log_access(LDAP_DEBUG_STATS,
-                             "conn=%" PRIu64 " %s %i-bit %s; missing subject\n",
-                             conn->c_connid,
-                             sslversion, keySize, cipher ? cipher : "NULL");
+            if (log_format != LOG_FORMAT_DEFAULT) {
+                slapd_log_pblock_init(&logpb, log_format, NULL);
+                logpb.conn_id = conn->c_connid;
+                logpb.msg = "missing subject";
+                logpb.tls_version = sslversion;
+                logpb.keysize = keySize;
+                logpb.cipher = cipher ? cipher : "NULL";
+                slapd_log_access_tls_client_auth(&logpb);
+            } else {
+                slapi_log_access(LDAP_DEBUG_STATS,
+                                "conn=%" PRIu64 " %s %i-bit %s; missing subject\n",
+                                conn->c_connid,
+                                sslversion, keySize, cipher ? cipher : "NULL");
+            }
             goto done;
-        }
-        {
+        } else {
             char *issuer = issuer_of(clientCert);
             char sbuf[BUFSIZ], ibuf[BUFSIZ];
             (void)slapi_getSSLVersion_str(channelInfo.protocolVersion,
                                           sslversion, sizeof(sslversion));
-            slapi_log_access(LDAP_DEBUG_STATS,
-                             "conn=%" PRIu64 " %s %i-bit %s; client %s; issuer %s\n",
-                             conn->c_connid,
-                             sslversion, keySize,
-                             cipher ? cipher : "NULL",
-                             escape_string(subject, sbuf),
-                             issuer ? escape_string(issuer, ibuf) : "NULL");
+            if (log_format != LOG_FORMAT_DEFAULT) {
+                slapd_log_pblock_init(&logpb, log_format, NULL);
+                logpb.conn_id = conn->c_connid;
+                logpb.tls_version = sslversion;
+                logpb.keysize = keySize;
+                logpb.cipher = cipher ? cipher : "NULL";
+                logpb.subject = escape_string(subject, sbuf);
+                logpb.issuer = issuer ? escape_string(issuer, ibuf) : "NULL";
+                slapd_log_access_tls_client_auth(&logpb);
+            } else {
+                slapi_log_access(LDAP_DEBUG_STATS,
+                        "conn=%" PRIu64 " %s %i-bit %s; client %s; issuer %s\n",
+                        conn->c_connid,
+                        sslversion, keySize,
+                        cipher ? cipher : "NULL",
+                        escape_string(subject, sbuf),
+                        issuer ? escape_string(issuer, ibuf) : "NULL");
+            }
             if (issuer)
-                free(issuer);
+                slapi_ch_free_string(&issuer);
         }
         slapi_dn_normalize(subject);
         {
             LDAPMessage *chain = NULL;
             char *basedn = config_get_basedn();
-            int err;
 
             err = ldapu_cert_to_ldap_entry(clientCert, internal_ld, basedn ? basedn : "" /*baseDN*/, &chain);
             if (err == LDAPU_SUCCESS && chain) {
@@ -505,18 +576,37 @@ handle_handshake_done(PRFileDesc *prfd, void *clientData)
         slapi_sdn_free(&sdn);
         (void)slapi_getSSLVersion_str(channelInfo.protocolVersion,
                                       sslversion, sizeof(sslversion));
-        slapi_log_access(LDAP_DEBUG_STATS,
-                         "conn=%" PRIu64 " %s client bound as %s\n",
-                         conn->c_connid,
-                         sslversion, clientDN);
+        if (log_format != LOG_FORMAT_DEFAULT) {
+            slapd_log_pblock_init(&logpb, log_format, NULL);
+            logpb.conn_id = conn->c_connid;
+            logpb.msg = "client bound";
+            logpb.tls_version = sslversion;
+            logpb.client_dn = clientDN;
+            slapd_log_access_tls_client_auth(&logpb);
+        } else {
+            slapi_log_access(LDAP_DEBUG_STATS,
+                            "conn=%" PRIu64 " %s client bound as %s\n",
+                            conn->c_connid,
+                            sslversion, clientDN);
+        }
     } else if (clientCert != NULL) {
         (void)slapi_getSSLVersion_str(channelInfo.protocolVersion,
                                       sslversion, sizeof(sslversion));
-        slapi_log_access(LDAP_DEBUG_STATS,
-                         "conn=%" PRIu64 " %s failed to map client "
-                         "certificate to LDAP DN (%s)\n",
-                         conn->c_connid,
-                         sslversion, extraErrorMsg);
+        if (log_format != LOG_FORMAT_DEFAULT) {
+            slapd_log_pblock_init(&logpb, log_format, NULL);
+            logpb.conn_id = conn->c_connid;
+            logpb.msg = "failed to map client certificate to LDAP DN";
+            logpb.tls_version = sslversion;
+            logpb.err = err;
+            logpb.err_str = extraErrorMsg;
+            slapd_log_access_tls_client_auth(&logpb);
+        } else {
+            slapi_log_access(LDAP_DEBUG_STATS,
+                            "conn=%" PRIu64 " %s failed to map client "
+                            "certificate to LDAP DN (%s)\n",
+                            conn->c_connid,
+                            sslversion, extraErrorMsg);
+        }
     }
 
     /*
diff --git a/ldap/servers/slapd/log.c b/ldap/servers/slapd/log.c
index f535011ab..91ba23047 100644
--- a/ldap/servers/slapd/log.c
+++ b/ldap/servers/slapd/log.c
@@ -7270,6 +7270,8 @@ slapd_log_pblock_init(slapd_log_pblock *logpb, int32_t log_format, Slapi_PBlock
         slapi_pblock_get(pb, SLAPI_CONNECTION, &conn);
     }
 
+    memset(logpb, 0, sizeof(slapd_log_pblock));
+
     logpb->loginfo = &loginfo;
     logpb->level = 256; /* default log level */
     logpb->log_format = log_format;
diff --git a/ldap/servers/slapd/slapi-private.h b/ldap/servers/slapd/slapi-private.h
index 6438a81fe..da232ae2f 100644
--- a/ldap/servers/slapd/slapi-private.h
+++ b/ldap/servers/slapd/slapi-private.h
@@ -1549,6 +1549,13 @@ typedef struct slapd_log_pblock {
     PRBool using_tls;
     PRBool haproxied;
     const char *bind_dn;
+    /* TLS */
+    const char *tls_version;
+    int32_t keysize;
+    const char *cipher;
+    const char *subject;
+    const char *issuer;
+    const char *client_dn;
     /* Close connection */
     const char *close_error;
     const char *close_reason;
@@ -1619,6 +1626,7 @@ typedef struct slapd_log_pblock {
     const char *oid;
     const char *msg;
     const char *name;
+    const char *err_str;
     LDAPControl **request_controls;
     LDAPControl **response_controls;
 } slapd_log_pblock;
@@ -1645,6 +1653,8 @@ int32_t slapd_log_access_entry(slapd_log_pblock *logpb);
 int32_t slapd_log_access_referral(slapd_log_pblock *logpb);
 int32_t slapd_log_access_extop(slapd_log_pblock *logpb);
 int32_t slapd_log_access_sort(slapd_log_pblock *logpb);
+int32_t slapd_log_access_tls(slapd_log_pblock *logpb);
+int32_t slapd_log_access_tls_client_auth(slapd_log_pblock *logpb);
 
 #ifdef __cplusplus
 }
-- 
2.49.0

